routemaster/models/batch.rb

```diff
@@ -1,56 +1,244 @@
 require 'routemaster/models'
 require 'routemaster/models/message'
+require 'routemaster/models/queue'
+require 'routemaster/models/subscriber'
+require 'routemaster/mixins/redis'
+require 'routemaster/mixins/assert'
 require 'routemaster/mixins/log'
+require 'routemaster/services/codec'
+require 'wisper'
 
 module Routemaster
   module Models
     # Abstracts an ordered list of Message
     class Batch
+      include Mixins::Redis
+      include Mixins::Assert
       include Mixins::Log
-      extend Forwardable
+      include Wisper::Publisher
 
-      def initialize(consumer)
-        @consumer = consumer
-        @batch = []
-        @monitor = Monitor.new
-      end
+      Inconsistency    = Class.new(RuntimeError)
 
-      delegate %i[push length] => :@batch
+      # number of prefix metdata items in a batch list
+      PREFIX_COUNT = 3
 
-      def synchronize(&block)
-        @monitor.synchronize { block.call(self) }
-        self
+
+      attr_reader :uid, :deadline
+
+
+      def initialize(uid:, deadline:nil, subscriber:nil)
+        @uid        = uid
+        @deadline   = deadline
+        @subscriber = subscriber
       end
 
-      def nack
-        synchronize do
-          @batch.each { |msg| @consumer.nack(msg) }
-          _flush
+
+      def subscriber
+        @subscriber ||= begin
+          return if _subscriber_name.nil?
+          Subscriber.find(_subscriber_name)
         end
-        self
       end
 
-      def ack  
-        synchronize do
-          @batch.each { |msg| @consumer.ack(msg) }
-          _flush
+
+      # Return the number of events in the batch (memoised)
+      def length
+        @_length ||= begin
+          raw = _redis.llen(_batch_key)
+          return if raw.nil? || raw.zero?
```

Comment:
If it's a 'broken' batch then yeah an exception sounds like it would make more sense.